File: microfreak/transport.py
"""The injected seam: byte movement and message-boundary reassembly, nothing
more.

Poll model, deliberately: the core stays single-threaded and thread-free;
push-style backends (rtmidi callback, CoreMIDI MIDIReadProc, ALSA events)
buffer into an internal queue inside the adapter (queue.Queue in Python, a
locked array + semaphore in Swift). Discovery is a per-backend factory
concern, not part of the interface.
"""
from __future__ import annotations

from typing import Optional

try:
    from typing import Protocol, runtime_checkable
except ImportError:  # pragma: no cover - 3.9+ always has these
    Protocol = object          # type: ignore

    def runtime_checkable(cls):  # type: ignore
        return cls


@runtime_checkable
class Transport(Protocol):
    """Structural interface every backend implements.

    Contract:
    - send() transmits one complete SysEx message F0..F7, atomically.
    - receive() returns the next complete inbound SysEx message, or None on
      timeout. Arrival order is preserved; the transport buffers and never
      drops. The transport's only jobs are byte movement and message-boundary
      reassembly — no parsing, filtering, matching, retries, or threading
      above the seam.
    - close() releases the backend; further calls may raise TransportError.
    """

    def send(self, message: bytes) -> None:
        """Send one complete SysEx message F0..F7, atomically."""
        ...

    def receive(self, timeout: float) -> Optional[bytes]:
        """Next complete inbound SysEx message, or None on timeout."""
        ...

    def close(self) -> None:
        ...

File: ios/FreakCore/Sources/FreakCore/Transport.swift
// Transport.swift — the injected seam: byte movement and message-boundary
// reassembly, nothing more (transport.py).
//
// Poll model, deliberately: the core stays single-threaded and thread-free;
// push-style backends (CoreMIDI receive blocks, rtmidi callbacks) buffer
// into an internal locked queue inside the adapter. Discovery is a
// per-backend factory concern, not part of the interface.
//
// Contract:
// - send() transmits one complete SysEx message F0..F7, atomically.
// - receive() returns the next complete inbound SysEx message, or nil on
//   timeout. Arrival order is preserved; the transport buffers and never
//   drops. The transport's only jobs are byte movement and message-boundary
//   reassembly — no parsing, filtering, matching, retries, or threading
//   above the seam.
// - close() releases the backend; further calls may throw FreakError.transport.
//
// Adapters catch every backend error and throw .transport / .deviceNotFound
// with the backend description in `detail`. Implementations are internally
// synchronized and declared `@unchecked Sendable` with a one-line
// justification comment at the declaration.

import Foundation

public protocol Transport: AnyObject, Sendable {
    /// Send one complete SysEx message F0..F7, atomically.
    func send(_ message: Data) throws
    /// Next complete inbound SysEx message, or nil on timeout.
    /// timeout <= 0 means "return immediately if nothing is queued".
    /// Blocks the calling thread up to `timeout` — callers run on the
    /// device actor's dedicated dispatch queue, never the cooperative pool.
    func receive(timeout: TimeInterval) throws -> Data?
    /// Release the backend; further calls may throw FreakError.transport.
    func close()
}
